File: frontend/src/utils/hostNetwork.ts
// Host-network handling shared by the standard (networkPolicyGenerator) and
// Cilium (ciliumPolicyGenerator) policy builders.
//
// A pod with `spec.hostNetwork: true` has no netns of its own: its IP IS the
// node IP, and the CNI sees its traffic as node (host) identity, not pod
// identity. Two consequences for generated policy:
//
//  1. A PEER that resolves to a host-network pod cannot be matched by a
//     podSelector — the packets carry the node's identity. The standard
//     generator emits an ipBlock for the observed node IP instead; the Cilium
//     generator emits `fromEntities`/`toEntities: [host, remote-node]` (both,
//     because the peer may sit on the local node or a remote one).
//  2. A TARGET that is host-network cannot be selected at all by a namespaced
//     NetworkPolicy / CiliumNetworkPolicy. The policy is still emitted, with a
//     leading warning block, so the user is not handed a silently inert rule.
//
// Text here must stay byte-identical to llm-bridge and the advisor: the
// fixture tests compare the emitted comment lines.

import type { PodInfo, PodNodeData } from '../types';

/** `pod_obj.spec.nodeName` when the stored manifest carries it. */
export function specNodeName(pod: PodInfo): string | undefined {
  const spec = pod.pod_obj?.spec as { nodeName?: unknown } | undefined;
  return typeof spec?.nodeName === 'string' && spec.nodeName ? spec.nodeName : undefined;
}

/** Cilium entities that cover a host-network peer on any node. */
export const HOST_NETWORK_ENTITIES: readonly string[] = ['host', 'remote-node'];

/** Which policy kind is being explained — drives the selector wording. */
export type HostNetworkPolicyKind = 'standard' | 'cilium';

const SELECTOR_WORD: Record<HostNetworkPolicyKind, string> = {
  standard: 'podSelector',
  cilium: 'endpointSelector',
};

/**
 * The `# ...` line placed above a rule whose peer is a host-network pod.
 * `name` is the owning workload when known, else the pod name.
 */
export function hostNetworkPeerComment(
  kind: HostNetworkPolicyKind,
  namespace: string,
  name: string,
  node: string,
): string {
  return `host-network peer ${namespace}/${name} on node ${node} — ${SELECTOR_WORD[kind]} cannot match host traffic`;
}

/**
 * The rule comment for a Service whose backing pods are host-network. The
 * Service is named `<ns>/svc/<name>`; the node is the sorted, deduplicated
 * node list of those backends (a ClusterIP has no single node), falling back
 * to the observed IP. Mirrors the advisor's hostNetworkServiceComment.
 */
export function hostNetworkServiceComment(
  kind: HostNetworkPolicyKind,
  namespace: string,
  svcName: string,
  backends: readonly PodInfo[],
  peerIP: string,
): string {
  const nodes = Array.from(
    new Set(backends.map((p) => p.node_name || specNodeName(p) || '').filter((n) => n !== '')),
  ).sort();
  const node = nodes.length > 0 ? nodes.join(',') : peerIP;
  return `host-network peer ${namespace}/svc/${svcName} on node ${node} — ${SELECTOR_WORD[kind]} cannot match host traffic`;
}

/** `labels` carries every key/value of `selector`. */
export function labelsContain(labels: Record<string, string> | undefined, selector: Record<string, string>): boolean {
  if (!labels) return false;
  return Object.entries(selector).every(([k, v]) => labels[k] === v);
}

/**
 * The host-network pods backing a Service: alive pods in the Service's
 * namespace whose labels contain the whole selector, sorted by name. Empty
 * when the Service has no selector, no such pods, or NONE of them is
 * host-network. `pods` null means the listing failed — treated as unknown,
 * i.e. the pre-existing podSelector rendering.
 */
export function hostNetworkServiceBackends(
  pods: readonly PodInfo[] | null,
  svcNamespace: string | undefined,
  selector: Record<string, string> | undefined,
): PodInfo[] {
  if (!pods || !svcNamespace || !selector || Object.keys(selector).length === 0) return [];
  return pods
    .filter((p) => !p.is_dead && p.pod_namespace === svcNamespace && p.host_network === true)
    .filter((p) => labelsContain(p.pod_obj?.metadata?.labels, selector))
    .sort((a, b) => (a.pod_name < b.pod_name ? -1 : a.pod_name > b.pod_name ? 1 : 0));
}

/**
 * Leading warning block for a policy whose TARGET workload is host-network.
 * Line breaks are fixed (not re-flowed) so the three generators emit the
 * same bytes.
 */
export function hostNetworkTargetWarning(kind: HostNetworkPolicyKind, namespace: string, workload: string): string[] {
  const selector = kind === 'cilium' ? 'CiliumNetworkPolicy endpointSelector' : 'NetworkPolicy podSelector';
  return [
    `WARNING: ${namespace}/${workload} runs with hostNetwork: true. A ${selector} cannot select`,
    'host-network pods; this policy will have no effect. Use a CiliumClusterwideNetworkPolicy',
    'with a nodeSelector (host firewall) instead.',
  ];
}

/** The name a host-network target is reported under: its workload when known. */
export function hostNetworkTargetName(pod: PodNodeData): string {
  return pod.pod.workload_name || pod.pod.pod_identity || pod.pod.pod_name;
}

/**
 * Whether the policy target runs with hostNetwork. An identity group may hold
 * several pods; any one of them being host-network means the podSelector is
 * inert for it. null/absent (old broker) is NOT host-network.
 *
 * The group-wide check is a frontend-only notion (advisor/llm-bridge take a
 * single pod): mid-rollout, when a workload's hostNetwork flag has flipped and
 * old and new pods coexist, the warning must still fire rather than silently
 * produce a policy that is ineffective for half the group.
 */
export function isHostNetworkTarget(pod: PodNodeData): boolean {
  if (pod.pod.host_network === true) return true;
  return Array.isArray(pod.pods) && pod.pods.some((p) => p?.host_network === true);
}

/** Render a list of comment lines as YAML `# ` lines at the given indent. */
export function yamlComments(lines: readonly string[] | undefined, indent = ''): string[] {
  if (!lines || lines.length === 0) return [];
  return lines.map((l) => `${indent}# ${l}`);
}

File: advisor/pkg/network/hostnetwork.go
package network

import (
	"fmt"
	"sort"
	"strings"

	"github.com/kguardian-dev/kguardian/advisor/pkg/api"
	"github.com/kguardian-dev/kguardian/advisor/pkg/common"
	log "github.com/rs/zerolog/log"
	"sigs.k8s.io/yaml"
)

// Host-network peers and targets.
//
// A pod running with spec.hostNetwork: true has no pod IP of its own — it
// shares the node's IP and, to the CNI, carries the node's identity. A
// NetworkPolicy podSelector (or Cilium endpointSelector) built from that pod's
// labels therefore never matches its traffic; the rule is silently
// ineffective. The generators handle two cases:
//
//   - a PEER resolves to a host-network pod: emit the observed node IP as an
//     ipBlock (NetworkPolicy) or the host + remote-node entities (Cilium)
//     instead of a label selector, and leave a YAML comment above the rule
//     saying why.
//   - the TARGET pod is itself host-network: the policy is emitted unchanged
//     but headed by a WARNING comment block — no selector can select it, and
//     the right tool is a CiliumClusterwideNetworkPolicy with a nodeSelector.
//
// The Kubernetes API types cannot carry comments, so the generators return
// them beside the policy in a PolicyComments and MarshalPolicyYAML splices
// them into the marshalled text. With no comments the output is byte-identical
// to yaml.Marshal, which is what keeps every pre-existing golden unchanged.

// PolicyComments is the set of YAML comments a generator wants rendered with a
// policy. Header lines precede the document; Ingress/Egress map a rule index
// (position in spec.ingress / spec.egress) to the comment lines rendered
// directly above that rule. Lines are stored without the leading "# ".
type PolicyComments struct {
	Header  []string
	Ingress map[int][]string
	Egress  map[int][]string
}

// CommentedPolicyGenerator is implemented by generators that can explain
// their output. PolicyService prefers it when available; plain
// PolicyGenerator implementations keep working unchanged.
type CommentedPolicyGenerator interface {
	PolicyGenerator
	// GenerateWithComments is Generate plus the comments to render with the
	// policy. The returned *PolicyComments may be nil.
	GenerateWithComments(podName string, podTraffic []api.PodTraffic, podDetail *api.PodDetail) (interface{}, *PolicyComments, error)
}

// IsEmpty reports whether there is nothing to render.
func (c *PolicyComments) IsEmpty() bool {
	return c == nil || (len(c.Header) == 0 && len(c.Ingress) == 0 && len(c.Egress) == 0)
}

func (c *PolicyComments) addHeader(lines ...string) {
	if c == nil {
		return
	}
	c.Header = append(c.Header, lines...)
}

func (c *PolicyComments) addIngress(idx int, line string) {
	if c == nil || line == "" {
		return
	}
	if c.Ingress == nil {
		c.Ingress = map[int][]string{}
	}
	c.Ingress[idx] = append(c.Ingress[idx], line)
}

func (c *PolicyComments) addEgress(idx int, line string) {
	if c == nil || line == "" {
		return
	}
	if c.Egress == nil {
		c.Egress = map[int][]string{}
	}
	c.Egress[idx] = append(c.Egress[idx], line)
}

func (c *PolicyComments) ruleLines(section string, idx int) []string {
	if c == nil {
		return nil
	}
	switch section {
	case "ingress":
		return c.Ingress[idx]
	case "egress":
		return c.Egress[idx]
	}
	return nil
}

// MarshalPolicyYAML marshals a generated policy exactly as yaml.Marshal does
// and then splices the comments in. A nil or empty PolicyComments yields the
// yaml.Marshal bytes untouched.
func MarshalPolicyYAML(policy interface{}, comments *PolicyComments) ([]byte, error) {
	out, err := yaml.Marshal(policy)
	if err != nil {
		return nil, err
	}
	return insertComments(out, comments), nil
}

// insertComments places Header lines before the document and each rule
// comment directly above its rule, at the rule's indentation.
//
// It works on the text sigs.k8s.io/yaml emits — two-space indentation, block
// sequences at the same column as their parent key — so a rule under
// `spec:`/`  egress:` starts with "  - ". Rules are counted in order within
// the section; the section ends at the next line at indent <= 2 that is not a
// rule start. Only the top-level `spec` key is inspected, so an `ingress` key
// anywhere else (a label, say) is never mistaken for the rule list.
func insertComments(doc []byte, c *PolicyComments) []byte {
	if c.IsEmpty() {
		return doc
	}
	text := strings.TrimRight(string(doc), "\n")
	lines := strings.Split(text, "\n")
	out := make([]string, 0, len(lines)+len(c.Header)+len(c.Ingress)+len(c.Egress))

	for _, h := range c.Header {
		out = append(out, "# "+h)
	}

	top, section := "", ""
	idx := -1
	for _, line := range lines {
		switch {
		case line != "" && line[0] != ' ':
			top = strings.TrimSuffix(line, ":")
			section = ""
		case top == "spec" && (line == "  ingress:" || line == "  egress:"):
			section = strings.TrimSpace(strings.TrimSuffix(line, ":"))
			idx = -1
		case section != "":
			if strings.HasPrefix(line, "  - ") {
				idx++
				for _, cm := range c.ruleLines(section, idx) {
					out = append(out, "  # "+cm)
				}
			} else if !strings.HasPrefix(line, "    ") {
				section = ""
			}
		}
		out = append(out, line)
	}
	return []byte(strings.Join(out, "\n") + "\n")
}

// isHostNetwork is true only when the broker positively reports
// host_network: true. nil (old broker / unknown) and false both mean "treat
// as a normal pod" — the pre-existing behaviour.
func isHostNetwork(d *api.PodDetail) bool {
	return d != nil && d.HostNetwork != nil && *d.HostNetwork
}

// hostWorkloadName names a host-network pod in comments: the owning workload
// when the broker knows it (stable across restarts), else the pod name.
func hostWorkloadName(d *api.PodDetail) string {
	if d.WorkloadName != "" {
		return d.WorkloadName
	}
	return d.Name
}

// hostNodeName names the node a host-network peer runs on: the broker's flat
// node_name, then the manifest's spec.nodeName, then the observed IP itself
// (which for a host-network pod IS the node address).
func hostNodeName(d *api.PodDetail, peerIP string) string {
	if d.NodeName != "" {
		return d.NodeName
	}
	if d.Pod.Spec.NodeName != "" {
		return d.Pod.Spec.NodeName
	}
	return peerIP
}

// hostNetworkPeerComment is the line rendered above a rule whose peer is a
// host-network pod. selector is the field the reader would otherwise expect
// ("podSelector" for NetworkPolicy, "endpointSelector" for Cilium).
func hostNetworkPeerComment(d *api.PodDetail, peerIP, selector string) string {
	return fmt.Sprintf("host-network peer %s/%s on node %s — %s cannot match host traffic",
		d.Namespace, hostWorkloadName(d), hostNodeName(d, peerIP), selector)
}

// hostNetworkTargetWarning is the header block for a policy whose target pod
// is itself host-network. kind/selector name the emitted resource and its
// selector field ("NetworkPolicy"/"podSelector", "CiliumNetworkPolicy"/
// "endpointSelector").
func hostNetworkTargetWarning(d *api.PodDetail, kind, selector string) []string {
	return []string{
		fmt.Sprintf("WARNING: %s/%s runs with hostNetwork: true. A %s %s cannot select",
			d.Namespace, hostWorkloadName(d), kind, selector),
		"host-network pods; this policy will have no effect. Use a CiliumClusterwideNetworkPolicy",
		"with a nodeSelector (host firewall) instead.",
	}
}

// hostNetworkServiceBackends returns the host-network pods backing svc: alive
// pods in the Service's namespace whose labels contain the whole selector,
// sorted by name. Empty when the Service has no selector, no backing pods, or
// none of them is host-network. A failed pod listing is logged and treated as
// "unknown" — the pre-existing podSelector rendering.
func hostNetworkServiceBackends(data BrokerData, svc *api.SvcDetail) []api.PodDetail {
	if svc == nil || len(svc.Service.Spec.Selector) == 0 || data == nil {
		return nil
	}
	pods, err := data.Pods()
	if err != nil {
		log.Warn().Err(err).Msgf("Cannot list pods to inspect backends of service %s/%s; assuming pod-network", svc.SvcNamespace, svc.SvcName)
		return nil
	}
	var out []api.PodDetail
	for _, p := range pods {
		if p.IsDead || p.Namespace != svc.SvcNamespace || !isHostNetwork(&p) {
			continue
		}
		if !labelsContain(p.Pod.Labels, svc.Service.Spec.Selector) {
			continue
		}
		out = append(out, p)
	}
	sort.Slice(out, func(i, j int) bool { return out[i].Name < out[j].Name })
	return out
}

func labelsContain(labels, selector map[string]string) bool {
	for k, v := range selector {
		if labels[k] != v {
			return false
		}
	}
	return true
}

// hostNetworkServiceComment is the rule comment for a Service whose backing
// pods are host-network. The Service is named "<ns>/svc/<name>"; the node is
// the sorted, deduplicated node list of the host-network backends (a
// ClusterIP has no single node), falling back to the observed IP.
func hostNetworkServiceComment(svc *api.SvcDetail, backends []api.PodDetail, peerIP, selector string) string {
	seen := map[string]bool{}
	var nodes []string
	for i := range backends {
		n := hostNodeName(&backends[i], "")
		if n == "" || seen[n] {
			continue
		}
		seen[n] = true
		nodes = append(nodes, n)
	}
	sort.Strings(nodes)
	node := strings.Join(nodes, ",")
	if node == "" {
		node = peerIP
	}
	return fmt.Sprintf("host-network peer %s/svc/%s on node %s — %s cannot match host traffic",
		svc.SvcNamespace, svc.SvcName, node, selector)
}

// hostNetworkServiceCIDRs is the NetworkPolicy peer set for a Service backed
// by host-network pods: one host CIDR per distinct backend pod IP (which IS a
// node IP), sorted bytewise on the raw IP like every other peer. NetworkPolicy
// is evaluated against the post-DNAT destination, so the ClusterIP itself is
// never the right peer. Unparseable IPs are skipped; an empty result means
// the caller drops the rule.
func hostNetworkServiceCIDRs(backends []api.PodDetail) []string {
	seen := map[string]bool{}
	var ips []string
	for _, b := range backends {
		if b.PodIP == "" || seen[b.PodIP] {
			continue
		}
		seen[b.PodIP] = true
		ips = append(ips, b.PodIP)
	}
	sort.Strings(ips)
	var cidrs []string
	for _, ip := range ips {
		cidr, err := common.HostCIDR(ip)
		if err != nil {
			log.Warn().Err(err).Msgf("Skipping host-network backend %s: cannot express it as a host CIDR", ip)
			continue
		}
		cidrs = append(cidrs, cidr)
	}
	return cidrs
}

// hostEntities is the Cilium entity list for a host-network peer. Both are
// needed: `host` is the local node, `remote-node` every other node, and the
// peer may sit on either.
var hostEntities = []string{"host", "remote-node"}
